src/server/routes/signIn.ts

```diff
@@ -15,84 +15,254 @@ import { ApiError } from '@/server/models/Error';
+    const { message, status } =
+      error instanceof ApiError ? error : new ApiError();
+
+    trackMetric('SignIn::Failure');
+
+    // re-render the sign in page on error, with pre-filled email
     const html = renderer('/signin', {
-      requestState: deepmerge(state, {
+      requestState: deepmerge(res.locals, {
+        globalMessage: {
+          error: message,
+        },
         pageData: {
           email,
         },
-        globalMessage: {
-          error: errorMessage,
-        },
       }),
       pageTitle: 'Sign in',
     });
-    res.type('html').send(html);
-  }),
-);
 
-router.post(
-  '/signin',
-  handleRecaptcha,
-  handleAsyncErrors(async (req: Request, res: ResponseWithRequestState) => {
-    const state = res.locals;
-
-    const { email = '' } = req.body;
-    const { password = '' } = req.body;
-
-    const { returnUrl } = state.pageData;
-
-    const { defaultReturnUri } = getConfiguration();
-
-    try {
-      const cookies = await authenticate(email, password, req.ip);
-      setIDAPICookies(res, cookies);
-    } catch (error) {
-      logger.error(`${req.method} ${req.originalUrl}  Error`, error);
-      const { message, status } =
-        error instanceof ApiError ? error : new ApiError();
-
-      trackMetric('SignIn::Failure');
-
-      // re-render the sign in page on error, with pre-filled email
-      const html = renderer('/signin', {
-        requestState: deepmerge(res.locals, {
-          globalMessage: {
-            error: message,
-          },
-          pageData: {
-            email,
-          },
-        }),
-        pageTitle: 'Sign in',
-      });
-
-      return res.status(status).type('html').send(html);
-    }
+    return res.status(status).type('html').send(html);
+  }
+};
 
-    trackMetric('SignIn::Success');
+const oktaSignInController = async (
+  req: Request,
+  res: ResponseWithRequestState,
+) => {
+  // get the email and password from the request body
+  const { email = '', password = '' } = req.body;
 
-    return res.redirect(303, returnUrl || defaultReturnUri);
-  }),
-);
+  try {
+    // attempt to authenticate with okta
+    // if authentication fails, it will fall through to the catch
+    // the response contains a one time use sessionToken that we can exchange
+    // for a session cookie
+    const response = await authenticateWithOkta({
+      username: email,
+      password,
+    });
+
+    // we're authenticated track this metric
+    trackMetric('OktaSignIn::Success');
+
+    // we now need to generate an okta session
+    // so we'll call the OIDC /authorize endpoint which sets a session cookie
+    // we'll pretty much be performing the Authorization Code Flow
+    return performAuthCodeFlow(res, response.sessionToken);
+  } catch (error) {
+    trackMetric('OktaSignIn::Failure');
+
+    logger.error('Okta authentication error:', error);
+
+    const { message, status } = ((error: unknown) => {
```

Comment:
could this function have a name for extra clarity perhaps?

Reply:
i'll refactor this out into it's own method!